### ISSUE TITLE
Wrapper is much slower for buffers than numpy arrays

### ISSUE THREAD

**Issue by seibert** [numba/numba#1257]
Passing a buffer to a nopython function is much slower than the equivalent numpy array:
```
@jit(nopython=True)
def noop(s):
    return 0

b = b'\0'*12
a = np.frombuffer(b, dtype=np.uint8)
%timeit noop(b)
%timeit noop(a)

----

The slowest run took 38.06 times longer than the fastest. This could mean that an intermediate result is being cached 
10000 loops, best of 3: 51.1 µs per loop
The slowest run took 84.82 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 270 ns per loop
```

**Comment by pitrou** [comment]
We have two avenues for making this faster:
* speed-up typeof() (which is really typing.Context.resolve_argument_type())
* add tons of caches in _dispatcher.c

**Comment by pitrou** [comment]
This is addressed by PR #1273.

**Comment by pitrou** [comment]
PR #1273 was merged, closing.
